musmart/algorithm/complexity.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""Implements various functions related to complexity and information density."""
+
+from typing import Iterable, Hashable, Union
+
+
+__all__ = ["lz77_size", "lz77_compress"]
+__author__ = "Huw Cheston"
+
+
+def lz77_compress(str_representation: str, window_size) -> list[tuple]:
+    """Runs the LZ77 compression algorithm over the input `data`, with given `window_size`"""
+    # TODO: this should use a list, rather than a string representation
+    compress_list = []
+    index = 0
+    while index < len(str_representation):
+        best_offset = -1
+        best_length = -1
+        best_match = ''
+        # Search for the longest match in the sliding window
+        for length in range(1, min(len(str_representation) - index, window_size)):
+            substring = str_representation[index:index + length]
+            offset = str_representation.rfind(substring, max(0, index - window_size), index)
+            if offset != -1 and length > best_length:
+                best_offset = index - offset
+                best_length = length
+                best_match = substring
+        if best_match:
+            # Add the (offset, length, next_character) tuple to the compressed data
+            compress_list.append((best_offset, best_length, str_representation[index + best_length]))
+            index += best_length + 1
+        else:
+            # No match found, add a zero-offset tuple
+            compress_list.append((0, 0, str_representation[index]))
+            index += 1
+    return compress_list
+
+
+def lz77_size(sequence: Iterable[Hashable], window_size: int = 64, proportional: bool = False) -> Union[int, float]:
+    """
+    Applies the LZ77 compression algorithm to a discrete `sequence` and returns the length of the compressed string.
+
+    Here we calculate the complexity of a discrete `sequence` (could be pitch classes, binned inter-onset intervals)
+    by converting it into a string representation, compressing using LZ77 [1], then calculating the length of the
+    compressed string as in [2]. Higher values mean that more information is required to represent the sequence,
+    indicating greater complexity.
+
+    Currently, only sequences with fewer than 96 unique elements are supported.
+
+    Parameters
+    ----------
+    sequence : np.ndarray
```

Comment:
The types in the docstring should match the type hints in the source code. Or you could just delete the former.